File: src/main/services/ragService.js
/**
 * RAG (Retrieval-Augmented Generation) Service
 * Handles cross-referencing, pattern detection, and tag suggestions
 */

const { v4: uuidv4 } = require('uuid');
const anthropicService = require('./anthropicService');

class RAGService {
    constructor(db, embeddingService) {
        this.db = db;
        this.embeddingService = embeddingService;
    }

    /**
     * Analyze an event and find cross-references with other events
     * @param {string} eventId - Event ID to analyze
     * @param {number} similarityThreshold - Minimum similarity score (0-1)
     * @returns {Promise<Object>} Analysis results with cross-references
     */
    async analyzeEventCrossReferences(eventId, similarityThreshold = 0.75) {
        try {
            // Get the source event
            const sourceEvent = this.db.prepare(`
                SELECT * FROM events WHERE event_id = ?
            `).get(eventId);

            if (!sourceEvent) {
                throw new Error('Event not found');
            }

            // Find similar events using embeddings
            const similarResult = await this.embeddingService.findSimilarEvents(
                eventId,
                similarityThreshold,
                20 // Get top 20 similar events
            );

            if (!similarResult.success || similarResult.similar_events.length === 0) {
                return {
                    success: true,
                    cross_references: [],
                    message: 'No similar events found'
                };
            }

            // Use LLM to analyze relationships
            const crossReferences = [];
            for (const similarEvent of similarResult.similar_events) {
                const analysis = await this._analyzePairRelationship(sourceEvent, similarEvent);
                if (analysis.hasRelationship) {
                    crossReferences.push({
                        event_id_2: similarEvent.event_id,
                        relationship_type: analysis.type,
                        confidence_score: analysis.confidence,
                        explanation: analysis.explanation,
                        similarity_score: similarEvent.similarity_score
                    });
                }
            }

            return {
                success: true,
                cross_references: crossReferences,
                source_event_id: eventId
            };
        } catch (error) {
            console.error('Error analyzing cross-references:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Analyze relationship between two events using LLM
     * @private
     */
    async _analyzePairRelationship(event1, event2) {
        if (!anthropicService.isInitialized()) {
            // Fallback to basic heuristics if LLM not available
            return this._basicRelationshipHeuristic(event1, event2);
        }

        const prompt = `Analyze the relationship between these two life events:

EVENT 1:
Title: ${event1.title}
Date: ${event1.start_date}${event1.end_date ? ` to ${event1.end_date}` : ''}
Description: ${event1.description || 'N/A'}
Category: ${event1.category || 'N/A'}

EVENT 2:
Title: ${event2.title}
Date: ${event2.start_date}${event2.end_date ? ` to ${event2.end_date}` : ''}
Description: ${event2.description || 'N/A'}
Category: ${event2.category || 'N/A'}

Determine if there is a meaningful relationship between these events. Return ONLY valid JSON:

{
  "hasRelationship": true/false,
  "type": "causal|thematic|temporal|person|location|other",
  "confidence": 0.85,
  "explanation": "Brief explanation of the relationship"
}

Relationship types:
- causal: Event 1 caused or led to Event 2
- thematic: Events share similar themes or topics
- temporal: Events are part of a sequence or pattern over time
- person: Events involve the same people
- location: Events occurred in the same place
- other: Other meaningful connection

Return false for hasRelationship if the events are not meaningfully connected.`;

        try {
            const response = await anthropicService.client.messages.create({
                model: 'claude-sonnet-4-20250514',
                max_tokens: 1000,
                temperature: 0.3,
                messages: [{
                    role: 'user',
                    content: prompt
                }]
            });

            const content = response.content[0].text;
            const jsonMatch = content.match(/\{[\s\S]*\}/);

            if (jsonMatch) {
                return JSON.parse(jsonMatch[0]);
            }

            return { hasRelationship: false };
        } catch (error) {
            console.error('Error analyzing relationship with LLM:', error);
            return this._basicRelationshipHeuristic(event1, event2);
        }
    }

    /**
     * Basic heuristic for relationship detection (fallback when LLM unavailable)
     * @private
     */
    _basicRelationshipHeuristic(event1, event2) {
        // Check for same category
        if (event1.category === event2.category && event1.category !== 'other') {
            return {
                hasRelationship: true,
                type: 'thematic',
                confidence: 0.6,
                explanation: `Both events are categorized as ${event1.category}`
            };
        }

        // Check for temporal proximity (within 30 days)
        const date1 = new Date(event1.start_date);
        const date2 = new Date(event2.start_date);
        const daysDiff = Math.abs((date2 - date1) / (1000 * 60 * 60 * 24));

        if (daysDiff <= 30) {
            return {
                hasRelationship: true,
                type: 'temporal',
                confidence: 0.5,
                explanation: `Events occurred within ${Math.round(daysDiff)} days of each other`
            };
        }

        return { hasRelationship: false };
    }

    /**
     * Store cross-reference in database
     * @param {string} eventId1 - First event ID
     * @param {string} eventId2 - Second event ID
     * @param {string} relationshipType - Type of relationship
     * @param {number} confidenceScore - Confidence score (0-1)
     * @param {string} explanation - Explanation of the relationship
     */
    storeCrossReference(eventId1, eventId2, relationshipType, confidenceScore, explanation) {
        // Ensure consistent ordering (event_id_1 < event_id_2)
        const [id1, id2] = eventId1 < eventId2 ? [eventId1, eventId2] : [eventId2, eventId1];

        const referenceId = uuidv4();
        const stmt = this.db.prepare(`
            INSERT OR REPLACE INTO cross_references (
                reference_id, event_id_1, event_id_2,
                relationship_type, confidence_score, analysis_details
            ) VALUES (?, ?, ?, ?, ?, ?)
        `);

        stmt.run(
            referenceId,
            id1,
            id2,
            relationshipType,
            confidenceScore,
            JSON.stringify({ explanation })
        );

        return referenceId;
    }

    /**
     * Get cross-references for an event
     * @param {string} eventId - Event ID
     * @returns {Array} Cross-references
     */
    getCrossReferences(eventId) {
        const stmt = this.db.prepare(`
            SELECT cr.*,
                   e1.title as event1_title, e1.start_date as event1_date,
                   e2.title as event2_title, e2.start_date as event2_date
            FROM cross_references cr
            INNER JOIN events e1 ON cr.event_id_1 = e1.event_id
            INNER JOIN events e2 ON cr.event_id_2 = e2.event_id
            WHERE cr.event_id_1 = ? OR cr.event_id_2 = ?
            ORDER BY cr.confidence_score DESC
        `);

        const results = stmt.all(eventId, eventId);

        return results.map(row => ({
            ...row,
            analysis_details: JSON.parse(row.analysis_details)
        }));
    }

    /**
     * Run full RAG analysis on entire timeline
     * @param {number} similarityThreshold - Minimum similarity threshold
     * @returns {Promise<Object>} Analysis results
     */
    async analyzeFullTimeline(similarityThreshold = 0.75) {
        try {
            console.log('Starting full timeline RAG analysis...');

            // Get all events with embeddings
            const events = this.db.prepare(`
                SELECT e.event_id FROM events e
                INNER JOIN event_embeddings em ON e.event_id = em.event_id
            `).all();

            let totalCrossRefs = 0;
            const errors = [];

            for (const event of events) {
                try {
                    const result = await this.analyzeEventCrossReferences(
                        event.event_id,
                        similarityThreshold
                    );

                    if (result.success) {
                        // Store cross-references
                        for (const crossRef of result.cross_references) {
                            this.storeCrossReference(
                                event.event_id,
                                crossRef.event_id_2,
                                crossRef.relationship_type,
                                crossRef.confidence_score,
                                crossRef.explanation
                            );
                            totalCrossRefs++;
                        }
                    }
                } catch (error) {
                    errors.push({ event_id: event.event_id, error: error.message });
                }

                // Add delay to avoid rate limiting
                await new Promise(resolve => setTimeout(resolve, 500));
            }

            console.log(`Timeline analysis complete. Found ${totalCrossRefs} cross-references.`);

            return {
                success: true,
                total_events: events.length,
                cross_references_found: totalCrossRefs,
                errors
            };
        } catch (error) {
            console.error('Error analyzing timeline:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Detect patterns in timeline (recurring events, cycles)
     * @returns {Promise<Object>} Detected patterns
     */
    async detectPatterns() {
        try {
            const patterns = [];

            // Pattern 1: Recurring categories
            const categoryPattern = this._detectCategoryPatterns();
            if (categoryPattern.length > 0) {
                patterns.push({
                    type: 'recurring_categories',
                    description: 'Events that occur repeatedly in certain categories',
                    patterns: categoryPattern
                });
            }

            // Pattern 2: Temporal clusters
            const temporalPattern = this._detectTemporalClusters();
            if (temporalPattern.length > 0) {
                patterns.push({
                    type: 'temporal_clusters',
                    description: 'Time periods with high event density',
                    patterns: temporalPattern
                });
            }

            // Pattern 3: Era transitions
            const transitionPattern = this._detectEraTransitions();
            if (transitionPattern.length > 0) {
                patterns.push({
                    type: 'era_transitions',
                    description: 'Significant transitions between life phases',
                    patterns: transitionPattern
                });
            }

            return {
                success: true,
                patterns
            };
        } catch (error) {
            console.error('Error detecting patterns:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Detect recurring category patterns
     * @private
     */
    _detectCategoryPatterns() {
        const stmt = this.db.prepare(`
            SELECT category, COUNT(*) as count
            FROM events
            WHERE category IS NOT NULL AND category != 'other'
            GROUP BY category
            HAVING count >= 3
            ORDER BY count DESC
        `);

        return stmt.all();
    }

    /**
     * Detect temporal event clusters
     * @private
     */
    _detectTemporalClusters() {
        // Find periods with high event density
        const stmt = this.db.prepare(`
            SELECT
                strftime('%Y-%m', start_date) as month,
                COUNT(*) as event_count
            FROM events
            GROUP BY month
            HAVING event_count >= 3
            ORDER BY event_count DESC
            LIMIT 10
        `);

        return stmt.all();
    }

    /**
     * Detect era transitions (milestone events near era boundaries)
     * @private
     */
    _detectEraTransitions() {
        const stmt = this.db.prepare(`
            SELECT
                e.event_id,
                e.title,
                e.start_date,
                e.category,
                er.name as era_name
            FROM events e
            INNER JOIN eras er ON e.era_id = er.era_id
            WHERE e.category IN ('milestone', 'achievement', 'challenge')
            ORDER BY e.start_date
        `);

        return stmt.all();
    }

    /**
     * Suggest tags for an event based on similar events
     * @param {string} eventId - Event ID
     * @param {number} limit - Maximum number of suggestions
     * @returns {Promise<Array>} Suggested tags with confidence scores
     */
    async suggestTags(eventId, limit = 5) {
        try {
            // Find similar events
            const similarResult = await this.embeddingService.findSimilarEvents(eventId, 0.7, 10);

            if (!similarResult.success || similarResult.similar_events.length === 0) {
                return { success: true, suggestions: [] };
            }

            // Get tags from similar events
            const similarEventIds = similarResult.similar_events.map(e => e.event_id);
            const placeholders = similarEventIds.map(() => '?').join(',');

            const stmt = this.db.prepare(`
                SELECT t.tag_name, COUNT(*) as frequency,
                       AVG(et.confidence_score) as avg_confidence
                FROM event_tags et
                INNER JOIN tags t ON et.tag_id = t.tag_id
                WHERE et.event_id IN (${placeholders})
                  AND et.event_id NOT IN (
                      SELECT event_id FROM event_tags WHERE event_id = ?
                  )
                GROUP BY t.tag_name
                ORDER BY frequency DESC, avg_confidence DESC
                LIMIT ?
            `);

            const suggestions = stmt.all(...similarEventIds, eventId, limit);

            return {
                success: true,
                suggestions: suggestions.map(s => ({
                    tag_name: s.tag_name,
                    confidence: Math.min(s.avg_confidence * (s.frequency / similarEventIds.length), 1.0),
                    frequency: s.frequency
                }))
            };
        } catch (error) {
            console.error('Error suggesting tags:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }
}

module.exports = RAGService;

File: windows-native/src/MemoryTimeline.Core/Services/RagService.cs
using Microsoft.Extensions.Logging;
using MemoryTimeline.Data.Models;
using MemoryTimeline.Data.Repositories;
using Microsoft.EntityFrameworkCore;
using System.Text.Json;

namespace MemoryTimeline.Core.Services;

/// <summary>
/// RAG (Retrieval-Augmented Generation) service implementation.
/// Creates a short-lived DbContext per operation via IDbContextFactory.
/// Cross-reference detection is heuristic (temporal proximity + shared
/// category, confidences derived from embedding similarity) and results are
/// persisted via <see cref="ICrossReferenceRepository"/>.
/// </summary>
public class RagService : IRagService
{
    private readonly IEmbeddingService _embeddingService;
    private readonly IEventService _eventService;
    private readonly IDbContextFactory<Data.AppDbContext> _contextFactory;
    private readonly ICrossReferenceRepository _crossReferenceRepository;
    private readonly ILogger<RagService> _logger;

    public RagService(
        IEmbeddingService embeddingService,
        IEventService eventService,
        IDbContextFactory<Data.AppDbContext> contextFactory,
        ICrossReferenceRepository crossReferenceRepository,
        ILogger<RagService> logger)
    {
        _embeddingService = embeddingService;
        _eventService = eventService;
        _contextFactory = contextFactory;
        _crossReferenceRepository = crossReferenceRepository;
        _logger = logger;
    }

    /// <summary>
    /// Finds similar events using embedding-based similarity search.
    /// </summary>
    public async Task<List<SimilarEvent>> FindSimilarEventsAsync(string eventId, int topK = 10, double threshold = 0.7)
    {
        try
        {
            _logger.LogInformation("Finding similar events for {EventId}", eventId);

            await using var dbContext = await _contextFactory.CreateDbContextAsync();

            // Get the source event and its embedding.
            // NOTE: queries must use the mapped EmbeddingVector column, never the
            // [NotMapped] Embedding alias (untranslatable -> runtime throw).
            var sourceEventEmbedding = await dbContext.EventEmbeddings
                .AsNoTracking()
                .FirstOrDefaultAsync(ee => ee.EventId == eventId);

            if (sourceEventEmbedding == null || string.IsNullOrWhiteSpace(sourceEventEmbedding.EmbeddingVector))
            {
                _logger.LogWarning("No embedding found for event {EventId}", eventId);
                return new List<SimilarEvent>();
            }

            var queryEmbedding = TryDeserializeEmbedding(sourceEventEmbedding.EmbeddingVector);
            if (queryEmbedding == null)
            {
                // Malformed source embedding: return empty results instead of
                // throwing so one corrupt row can't break the Connections page.
                _logger.LogWarning(
                    "Stored embedding for event {EventId} is malformed; returning no similar events. " +
                    "Regenerate embeddings to fix.", eventId);
                return new List<SimilarEvent>();
            }

            // Get all other event embeddings
            var allEmbeddings = await dbContext.EventEmbeddings
                .AsNoTracking()
                .Where(ee => ee.EventId != eventId && ee.EmbeddingVector != null && ee.EmbeddingVector != "")
                .ToListAsync();

            var candidateEmbeddings = allEmbeddings
                .Select(ee => (ee.EventId, Embedding: TryDeserializeEmbedding(ee.EmbeddingVector)))
                .Where(x => x.Embedding != null)
                .Select(x => (x.EventId, x.Embedding!))
                .ToList();

            // Find K nearest neighbors
            var similarities = _embeddingService.FindKNearestNeighbors(
                queryEmbedding,
                candidateEmbeddings,
                topK,
                threshold);

            // Get event details
            var similarEvents = new List<SimilarEvent>();
            foreach (var similarity in similarities)
            {
                var eventEntity = await _eventService.GetEventByIdAsync(similarity.Id);
                if (eventEntity != null)
                {
                    similarEvents.Add(new SimilarEvent
                    {
                        EventId = eventEntity.EventId,
                        Title = eventEntity.Title,
                        Description = eventEntity.Description,
                        StartDate = eventEntity.StartDate,
                        Similarity = similarity.Similarity
                    });
                }
            }

            _logger.LogInformation("Found {Count} similar events for {EventId}", similarEvents.Count, eventId);
            return similarEvents;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error finding similar events");
            throw;
        }
    }

    /// <summary>
    /// Detects cross-references between events.
    /// Detection is HEURISTIC (temporal proximity and shared category, with
    /// confidences derived from embedding similarity where available) — no LLM
    /// call is made. New detections are persisted via the cross-reference
    /// repository (deduplicated per event pair), and previously stored
    /// references for the event are merged into the returned results.
    /// </summary>
    public async Task<List<CrossReferenceResult>> DetectCrossReferencesAsync(string eventId, IEnumerable<string>? candidateEventIds = null)
    {
        try
        {
            _logger.LogInformation("Detecting cross-references for event {EventId}", eventId);

            var sourceEvent = await _eventService.GetEventWithDetailsAsync(eventId);
            if (sourceEvent == null)
            {
                throw new Exception($"Event {eventId} not found");
            }

            // Get candidate events (either specified, or embedding-similar events).
            // similarityByEventId feeds the heuristic confidence scores.
            var candidates = new List<Event>();
            var similarityByEventId = new Dictionary<string, double>();

            if (candidateEventIds != null && candidateEventIds.Any())
            {
                foreach (var id in candidateEventIds)
                {
                    var evt = await _eventService.GetEventWithDetailsAsync(id);
                    if (evt != null) candidates.Add(evt);
                }
            }
            else
            {
                var similarEvents = await FindSimilarEventsAsync(eventId, topK: 20, threshold: 0.6);
                foreach (var similar in similarEvents)
                {
                    var evt = await _eventService.GetEventWithDetailsAsync(similar.EventId);
                    if (evt != null)
                    {
                        candidates.Add(evt);
                        similarityByEventId[similar.EventId] = similar.Similarity;
                    }
                }
            }

            // Start from what is already stored so previous detections are not lost
            var results = new List<CrossReferenceResult>();
            var knownPairs = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

            var storedReferences = await _crossReferenceRepository.GetReferencesForEventAsync(eventId);
            foreach (var stored in storedReferences)
            {
                results.Add(ToResult(stored, eventId));
                knownPairs.Add(PairKey(stored.EventId1, stored.EventId2));
            }

            // Heuristic detection over the candidates, persisting new pairs
            foreach (var candidate in candidates)
            {
                if (candidate.EventId == eventId)
                    continue;

                if (knownPairs.Contains(PairKey(eventId, candidate.EventId)))
                    continue;

                double? similarity = similarityByEventId.TryGetValue(candidate.EventId, out var s) ? s : null;
                var relationship = DetermineRelationshipHeuristically(sourceEvent, candidate, similarity);
                if (relationship == null)
                    continue;

                knownPairs.Add(PairKey(eventId, candidate.EventId));
                results.Add(relationship);

                // Persist (dedupe against concurrent writers via GetReferenceAsync).
                // Persistence failure must not break detection results.
                try
                {
                    var existing = await _crossReferenceRepository.GetReferenceAsync(eventId, candidate.EventId);
                    if (existing == null)
                    {
                        await _crossReferenceRepository.AddAsync(new CrossReference
                        {
                            ReferenceId = Guid.NewGuid().ToString(),
                            EventId1 = relationship.SourceEventId,
                            EventId2 = relationship.TargetEventId,
                            RelationshipType = ToStorageType(relationship.RelationshipType),
                            ConfidenceScore = relationship.Confidence,
                            AnalysisDetails = relationship.Reasoning,
                            CreatedAt = DateTime.UtcNow
                        });
                    }
                }
                catch (Exception persistEx)
                {
                    _logger.LogWarning(persistEx,
                        "Failed to persist cross-reference {Source} -> {Target}",
                        relationship.SourceEventId, relationship.TargetEventId);
                }
            }

            _logger.LogInformation("Detected {Count} cross-references for event {EventId} ({Stored} previously stored)",
                results.Count, eventId, storedReferences.Count());

            return results;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting cross-references");
            throw;
        }
    }

    /// <summary>
    /// Detects patterns in events.
    /// </summary>
    public async Task<PatternAnalysisResult> DetectPatternsAsync(DateTime? startDate = null, DateTime? endDate = null)
    {
        try
        {
            _logger.LogInformation("Detecting patterns in events");

            var result = new PatternAnalysisResult();

            // Get events in date range
            var events = startDate.HasValue && endDate.HasValue
                ? await _eventService.GetEventsByDateRangeAsync(startDate.Value, endDate.Value)
                : await _eventService.GetAllEventsAsync();

            var eventList = events.ToList();

            if (eventList.Count == 0)
            {
                return result;
            }

            // Detect category patterns
            result.CategoryPatterns = DetectCategoryPatterns(eventList);

            // Detect temporal clusters
            result.TemporalClusters = DetectTemporalClusters(eventList);

            // Detect era transitions
            result.EraTransitions = DetectEraTransitions(eventList);

            _logger.LogInformation("Detected {CategoryPatterns} category patterns, {TemporalClusters} temporal clusters, {EraTransitions} era transitions",
                result.CategoryPatterns.Count, result.TemporalClusters.Count, result.EraTransitions.Count);

            return result;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting patterns");
            throw;
        }
    }

    /// <summary>
    /// Suggests tags for an event based on similar events.
    /// </summary>
    public async Task<List<TagSuggestion>> SuggestTagsAsync(string eventId, int maxSuggestions = 5)
    {
        try
        {
            _logger.LogInformation("Suggesting tags for event {EventId}", eventId);

            // Find similar events
            var similarEvents = await FindSimilarEventsAsync(eventId, topK: 10, threshold: 0.75);

            if (similarEvents.Count == 0)
            {
                return new List<TagSuggestion>();
            }

            // Get tags from similar events with frequency counting
            var tagFrequency = new Dictionary<string, (int count, double totalSimilarity)>();

            foreach (var similar in similarEvents)
            {
                var eventTags = await _eventService.GetEventTagsAsync(similar.EventId);

                foreach (var tag in eventTags)
                {
                    if (tagFrequency.ContainsKey(tag.Name))
                    {
                        var current = tagFrequency[tag.Name];
                        tagFrequency[tag.Name] = (current.count + 1, current.totalSimilarity + similar.Similarity);
                    }
                    else
                    {
                        tagFrequency[tag.Name] = (1, similar.Similarity);
                    }
                }
            }

            // Calculate confidence scores and create suggestions
            var suggestions = tagFrequency
                .Select(kvp => new TagSuggestion
                {
                    TagName = kvp.Key,
                    Confidence = (kvp.Value.totalSimilarity / similarEvents.Count) * (kvp.Value.count / (double)similarEvents.Count),
                    Reason = $"Found in {kvp.Value.count} similar events"
                })
                .OrderByDescending(s => s.Confidence)
                .Take(maxSuggestions)
                .ToList();

            _logger.LogInformation("Generated {Count} tag suggestions for event {EventId}", suggestions.Count, eventId);
            return suggestions;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error suggesting tags");
            throw;
        }
    }

    /// <summary>
    /// Suggests tags for text before event creation.
    /// </summary>
    public async Task<List<TagSuggestion>> SuggestTagsForTextAsync(string title, string? description, int maxSuggestions = 5)
    {
        try
        {
            _logger.LogInformation("Suggesting tags for text: {Title}", title);

            // Generate embedding for the text
            var text = string.IsNullOrWhiteSpace(description) ? title : $"{title}. {description}";
            var queryEmbedding = await _embeddingService.GenerateEmbeddingAsync(text);

            // Get all event embeddings (mapped EmbeddingVector column — the
            // [NotMapped] Embedding alias is untranslatable in EF queries)
            await using var dbContext = await _contextFactory.CreateDbContextAsync();

            var allEmbeddings = await dbContext.EventEmbeddings
                .AsNoTracking()
                .Where(ee => ee.EmbeddingVector != null && ee.EmbeddingVector != "")
                .ToListAsync();

            var candidateEmbeddings = allEmbeddings
                .Select(ee => (ee.EventId, Embedding: TryDeserializeEmbedding(ee.EmbeddingVector)))
                .Where(x => x.Embedding != null)
                .Select(x => (x.EventId, x.Embedding!))
                .ToList();

            // Find similar events
            var similarities = _embeddingService.FindKNearestNeighbors(
                queryEmbedding,
                candidateEmbeddings,
                10,
                0.7);

            // Get tags from similar events
            var tagFrequency = new Dictionary<string, (int count, double totalSimilarity)>();

            foreach (var similarity in similarities)
            {
                var eventTags = await _eventService.GetEventTagsAsync(similarity.Id);

                foreach (var tag in eventTags)
                {
                    if (tagFrequency.ContainsKey(tag.Name))
                    {
                        var current = tagFrequency[tag.Name];
                        tagFrequency[tag.Name] = (current.count + 1, current.totalSimilarity + similarity.Similarity);
                    }
                    else
                    {
                        tagFrequency[tag.Name] = (1, similarity.Similarity);
                    }
                }
            }

            // Calculate confidence and create suggestions
            var suggestions = tagFrequency
                .Select(kvp => new TagSuggestion
                {
                    TagName = kvp.Key,
                    Confidence = (kvp.Value.totalSimilarity / similarities.Count) * (kvp.Value.count / (double)similarities.Count),
                    Reason = $"Found in {kvp.Value.count} similar events"
                })
                .OrderByDescending(s => s.Confidence)
                .Take(maxSuggestions)
                .ToList();

            _logger.LogInformation("Generated {Count} tag suggestions for text", suggestions.Count);
            return suggestions;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error suggesting tags for text");
            throw;
        }
    }

    #region Private Methods

    /// <summary>
    /// Safely deserializes a stored embedding, returning null (and logging) on malformed data
    /// instead of throwing so a single corrupt row cannot abort an entire similarity search.
    /// </summary>
    private float[]? TryDeserializeEmbedding(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
            return null;

        try
        {
            return JsonSerializer.Deserialize<float[]>(json);
        }
        catch (JsonException ex)
        {
            _logger.LogWarning(ex, "Failed to deserialize stored embedding; skipping");
            return null;
        }
    }

    /// <summary>
    /// Heuristic relationship detection with honest labelling: confidences are
    /// derived from embedding similarity and temporal proximity — never
    /// hard-coded — and the reasoning states that the result is a heuristic.
    /// </summary>
    private CrossReferenceResult? DetermineRelationshipHeuristically(Event source, Event target, double? similarity)
    {
        var daysDifference = Math.Abs((target.StartDate - source.StartDate).TotalDays);

        if (daysDifference <= 7)
        {
            // Proximity score: 1.0 (same day) down to ~0.0 (7 days apart)
            var proximity = 1.0 - (daysDifference / 7.0);

            // Blend with embedding similarity when we have one; otherwise the
            // proximity alone (discounted — we know less about the pair).
            var confidence = similarity.HasValue
                ? Math.Clamp(0.5 * similarity.Value + 0.5 * proximity, 0.0, 1.0)
                : Math.Clamp(0.6 * proximity, 0.0, 1.0);

            return new CrossReferenceResult
            {
                SourceEventId = source.EventId,
                TargetEventId = target.EventId,
                RelationshipType = CrossReferenceType.Temporal,
                Confidence = Math.Round(confidence, 3),
                Reasoning = similarity.HasValue
                    ? $"Heuristic: events occurred within {daysDifference:0.#} days of each other (embedding similarity {similarity.Value:0.00})"
                    : $"Heuristic: events occurred within {daysDifference:0.#} days of each other"
            };
        }

        if (!string.IsNullOrWhiteSpace(source.Category) &&
            string.Equals(source.Category, target.Category, StringComparison.OrdinalIgnoreCase))
        {
            // Thematic closeness is best measured by embedding similarity; the
            // shared category alone is a weak signal, reflected in the score.
            var confidence = similarity.HasValue
                ? Math.Clamp(similarity.Value, 0.0, 1.0)
                : 0.4;

            return new CrossReferenceResult
            {
                SourceEventId = source.EventId,
                TargetEventId = target.EventId,
                RelationshipType = CrossReferenceType.Thematic,
                Confidence = Math.Round(confidence, 3),
                Reasoning = similarity.HasValue
                    ? $"Heuristic: both events share the '{source.Category}' category (embedding similarity {similarity.Value:0.00})"
                    : $"Heuristic: both events share the '{source.Category}' category (no embedding similarity available)"
            };
        }

        return null;
    }

    /// <summary>Canonical unordered key for an event pair.</summary>
    private static string PairKey(string eventId1, string eventId2)
    {
        return string.CompareOrdinal(eventId1, eventId2) <= 0
            ? $"{eventId1}|{eventId2}"
            : $"{eventId2}|{eventId1}";
    }

    /// <summary>Converts a stored cross-reference row into a result, oriented from the requested event.</summary>
    private static CrossReferenceResult ToResult(CrossReference stored, string sourceEventId)
    {
        var isForward = string.Equals(stored.EventId1, sourceEventId, StringComparison.OrdinalIgnoreCase);

        return new CrossReferenceResult
        {
            SourceEventId = isForward ? stored.EventId1 : stored.EventId2,
            TargetEventId = isForward ? stored.EventId2 : stored.EventId1,
            RelationshipType = FromStorageType(stored.RelationshipType),
            Confidence = stored.ConfidenceScore ?? 0.0,
            Reasoning = stored.AnalysisDetails
        };
    }

    /// <summary>Maps the service-level relationship enum to the stored string value.</summary>
    private static string ToStorageType(CrossReferenceType type)
    {
        return type switch
        {
            CrossReferenceType.Causal => RelationshipType.Causal.ToStringValue(),
            CrossReferenceType.Thematic => RelationshipType.Thematic.ToStringValue(),
            CrossReferenceType.Temporal => RelationshipType.Temporal.ToStringValue(),
            CrossReferenceType.Participatory => RelationshipType.Person.ToStringValue(),
            CrossReferenceType.Locational => RelationshipType.Location.ToStringValue(),
            CrossReferenceType.Consequential => RelationshipType.Causal.ToStringValue(),
            _ => RelationshipType.Other.ToStringValue()
        };
    }

    /// <summary>Maps a stored relationship string back to the service-level enum.</summary>
    private static CrossReferenceType FromStorageType(string? storedType)
    {
        return RelationshipTypeExtensions.FromString(storedType ?? string.Empty) switch
        {
            RelationshipType.Causal => CrossReferenceType.Causal,
            RelationshipType.Temporal => CrossReferenceType.Temporal,
            RelationshipType.Person => CrossReferenceType.Participatory,
            RelationshipType.Location => CrossReferenceType.Locational,
            _ => CrossReferenceType.Thematic
        };
    }

    private List<CategoryPattern> DetectCategoryPatterns(List<Event> events)
    {
        var patterns = new List<CategoryPattern>();

        var categoryGroups = events.GroupBy(e => e.Category);

        foreach (var group in categoryGroups)
        {
            var pattern = new CategoryPattern
            {
                Category = group.Key ?? "Unknown",
                EventCount = group.Count()
            };

            // Calculate frequency (events per month)
            if (events.Count > 0)
            {
                var dateRange = events.Max(e => e.StartDate) - events.Min(e => e.StartDate);
                var months = Math.Max(1, dateRange.TotalDays / 30);
                pattern.Frequency = group.Count() / months;
            }

            patterns.Add(pattern);
        }

        return patterns.OrderByDescending(p => p.EventCount).ToList();
    }

    private List<TemporalCluster> DetectTemporalClusters(List<Event> events)
    {
        var clusters = new List<TemporalCluster>();
        var sortedEvents = events.OrderBy(e => e.StartDate).ToList();

        // Simple clustering: group events within 30 days
        for (int i = 0; i < sortedEvents.Count; i++)
        {
            var clusterEvents = new List<Event> { sortedEvents[i] };
            var startDate = sortedEvents[i].StartDate;

            for (int j = i + 1; j < sortedEvents.Count; j++)
            {
                if ((sortedEvents[j].StartDate - startDate).TotalDays <= 30)
                {
                    clusterEvents.Add(sortedEvents[j]);
                    i = j;
                }
                else
                {
                    break;
                }
            }

            if (clusterEvents.Count >= 3)  // Only clusters with 3+ events
            {
                clusters.Add(new TemporalCluster
                {
                    StartDate = clusterEvents.First().StartDate,
                    EndDate = clusterEvents.Last().StartDate,
                    EventCount = clusterEvents.Count,
                    EventIds = clusterEvents.Select(e => e.EventId).ToList(),
                    Theme = DetermineClusterTheme(clusterEvents)
                });
            }
        }

        return clusters;
    }

    private List<EraTransition> DetectEraTransitions(List<Event> events)
    {
        var transitions = new List<EraTransition>();

        // Simple transition detection based on category changes
        var sortedEvents = events.OrderBy(e => e.StartDate).ToList();

        for (int i = 1; i < sortedEvents.Count - 1; i++)
        {
            var prevCategory = sortedEvents[i - 1].Category;
            var currentCategory = sortedEvents[i].Category;

            if (prevCategory != currentCategory)
            {
                // Check if this is a sustained change (next few events are also different)
                var nextEvents = sortedEvents.Skip(i).Take(3).ToList();
                if (nextEvents.Count(e => e.Category == currentCategory) >= 2)
                {
                    transitions.Add(new EraTransition
                    {
                        TransitionDate = sortedEvents[i].StartDate,
                        FromTheme = prevCategory ?? "Unknown",
                        ToTheme = currentCategory ?? "Unknown",
                        Description = $"Transition from {prevCategory ?? "Unknown"} to {currentCategory ?? "Unknown"} focus"
                    });
                }
            }
        }

        return transitions;
    }

    private string? DetermineClusterTheme(List<Event> events)
    {
        // Find most common category in cluster
        var commonCategory = events
            .GroupBy(e => e.Category)
            .OrderByDescending(g => g.Count())
            .First()
            .Key;

        return commonCategory;
    }

    #endregion
}
